packages/backend/src/models/ProjectModel/ProjectModel.ts

```diff
@@ -383,4 +388,55 @@ export class ProjectModel {
             })
             .where('project_uuid', projectUuid);
     }
+
+    async getCacheExplores(
+        projectUuid: string,
+    ): Promise<(Explore | ExploreError)[]> {
+        const explores = await this.database(CacheExploresTableName)
+            .select(['explores'])
+            .where('project_uuid', projectUuid)
+            .limit(1);
+        if (explores.length > 0) return explores[0].explores;
+        return [];
+    }
+
+    async saveCacheExplores(
+        projectUuid: string,
+        explores: (Explore | ExploreError)[],
+    ): Promise<void> {
+        this.database.transaction(async (trx) => {
+            await trx(CacheExploresTableName)
+                .where('project_uuid', projectUuid)
+                .delete();
+
+            const result = await trx(CacheExploresTableName).insert({
+                project_uuid: projectUuid,
+                explores: JSON.stringify(explores),
+            });
+        });
+    }
+
+    async lockProcess(projectUuid: string, func: () => void): Promise<void> {
+        this.database.transaction(async (trx) => {
+            // pg_advisory_xact_lock takes a 64bit integer as key
+            // we can't use project_uuid (uuidv4) as key, not even a hash,
+            // so we will be using autoinc project_id from DB.
+            const projects = await this.database('projects')
+                .where('project_uuid', projectUuid)
+                .select('project_id')
+                .limit(1);
+            const projectId = projects[0].project_id;
+
+            // usage: pg_try_advisory_xact_lock(namespace, key)
+            // for projects we will be using namespace=1
+            const rawLock = await this.database.raw(
+                `SELECT pg_try_advisory_xact_lock('1', ${projectId});`,
+            );
+            const adquiresLock = rawLock.rows[0].pg_try_advisory_xact_lock;
+            if (!adquiresLock) return; // Lock is taken by another process, exiting
+
+            await trx.raw(`SELECT pg_advisory_xact_lock('1', ${projectId});`);
```

Comment:
I don't think you need to acquire the lock explicitly here

`pg_try_advisory_xact_lock` will return `true` if it successfully gets the lock, so if it returns `true` do work and return, if it returns `false` bail out.

Reply:
mmm, I think you are right, not sure why this didn't work until I set `await` on the `pg_advisory_xact_lock` method... anyway , it works now